### PR TITLE
Bump to xDSL supporting inplace bufferization

### PR DESCRIPTION
**Pull request by PapyChacal**
Just bump to xDSL supporting inplace bufferization and support inplace

Stacked on:
- #118

### CODE REVIEW

File: tests/test_xdsl_op_correctness.py [hunk 1]
```diff
@@ -131,8 +131,6 @@ def test_symbol_I():
     assert type(ops[0] == Return)
 
 
-# This test should fail, as we are trying to use an inplace operation
-@pytest.mark.xfail(reason="Cannot store to a field that is loaded from")
 def test_inplace():
     # Define a simple Devito Operator
     grid = Grid(shape=(3, 3))
```
Comment (PapyChacal) [inline tests/test_xdsl_op_correctness.py:136]: We probably want to check correct computations here then :thinking:

**Comment by PapyChacal** [comment]
Self-note: There's an issue with an unlowered memef.alloc
Which might mean the bufferization is not acting as expected, as I wouldn't expect an allocation in the in place case

**Comment by PapyChacal** [comment]
Okay so, Devito here is doing that (turning on OpenMP for clarity):
```c
for (int x = x_M; x >= x_m; x -= 1)
    {
      #pragma omp parallel num_threads(nthreads)
      {
        #pragma omp for schedule(static,1)
        for (int y = y_m; y <= y_M; y += 1)
        {
          u[t0][x + 1][y + 1] = r0*(-u[t0][x + 1][y + 1]) + r0*u[t0][x + 2][y + 1];
        }
      }
    }
```
So it sees there's a dependency only on `y` and expresses the computation to be sequential over `x` so that it can indeed do this one in place.
This is just not something the stencil dialect can express at this point :smiling_face_with_tear: 
@georgebisbas There are multiple ways from here onwards, let's discuss it when you have time. But indeed I don't see how to manage this without an extra allocation in xDSL in the time I have on my side

**Comment by georgebisbas** [comment]
> Okay so, Devito here is doing that (turning on OpenMP for clarity):
> 
> ```c
> for (int x = x_M; x >= x_m; x -= 1)
>     {
>       #pragma omp parallel num_threads(nthreads)
>       {
>         #pragma omp for schedule(static,1)
>         for (int y = y_m; y <= y_M; y += 1)
>         {
>           u[t0][x + 1][y + 1] = r0*(-u[t0][x + 1][y + 1]) + r0*u[t0][x + 2][y + 1];
>         }
>       }
>     }
> ```
> 
> So it sees there's a dependency only on `y` and expresses the computation to be sequential over `x` so that it can indeed do this one in place. This is just not something the stencil dialect can express at this point 🥲 @georgebisbas There are multiple ways from here onwards, let's discuss it when you have time. But indeed I don't see how to manage this without an extra allocation in xDSL in the time I have on my side

Thanks for that @PapyChacal , does xdsl produce one level of parallelism + correct results?
Let's discuss this yes.

File: tests/test_xdsl_op_correctness.py [hunk 2]
```diff
@@ -131,9 +131,23 @@ def test_symbol_I():
...
+    grid = Grid(shape=(3, 3))
+    u = TimeFunction(name="u", grid=grid, time_order=2)
+
+    u.data[:] = 0.0001
+
+    eq0 = Eq(u, u + 2)
+
+    xdsl_op = Operator([eq0], opt="xdsl")
+    xdsl_op.apply(time_M=5, dt=0.1)
+
```
Comment (georgebisbas) [inline tests/test_xdsl_op_correctness.py:145]: add tests remider

File: tests/test_xdsl_op_correctness.py [hunk 3]
```diff
@@ -131,9 +131,23 @@ def test_symbol_I():
...
+
+# This test should fail, as we are trying to use an inplace operation with some
+# dependencies
 @pytest.mark.xfail(reason="Cannot store to a field that is loaded from")
-def test_inplace():
+def test_inplace_II():
```
Comment (georgebisbas) [inline tests/test_xdsl_op_correctness.py:150]: probably no meaning for this

**Comment by PapyChacal** [comment]
> > Okay so, Devito here is doing that (turning on OpenMP for clarity):
> > ```c
> > for (int x = x_M; x >= x_m; x -= 1)
> >     {
> >       #pragma omp parallel num_threads(nthreads)
> >       {
> >         #pragma omp for schedule(static,1)
> >         for (int y = y_m; y <= y_M; y += 1)
> >         {
> >           u[t0][x + 1][y + 1] = r0*(-u[t0][x + 1][y + 1]) + r0*u[t0][x + 2][y + 1];
> >         }
> >       }
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So it sees there's a dependency only on `y` and expresses the computation to be sequential over `x` so that it can indeed do this one in place. This is just not something the stencil dialect can express at this point 🥲 @georgebisbas There are multiple ways from here onwards, let's discuss it when you have time. But indeed I don't see how to manage this without an extra allocation in xDSL in the time I have on my side
> 
> Thanks for that @PapyChacal , does xdsl produce one level of parallelism + correct results? Let's discuss this yes.

For the record, this does not seem to happen in TTI

**Pull request merged**